src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HMACCommon.cs

```diff
@@ -57,12 +57,33 @@ public void ChangeKey(byte[] key)
             if (key.Length > _blockSize && _blockSize > 0)
             {
                 // Perform RFC 2104, section 2 key adjustment.
-                if (_lazyHashProvider == null)
+
+                switch (_hashAlgorithmId)
                 {
-                    _lazyHashProvider = HashProviderDispenser.CreateHashProvider(_hashAlgorithmId);
+                    case HashAlgorithmNames.SHA256:
+                        modifiedKey = SHA256.HashData(key);
+                        break;
+                    case HashAlgorithmNames.SHA384:
+                        modifiedKey = SHA384.HashData(key);
+                        break;
+                    case HashAlgorithmNames.SHA512:
+                        modifiedKey = SHA512.HashData(key);
+                        break;
+                    case HashAlgorithmNames.SHA1:
+                        modifiedKey = SHA1.HashData(key);
+                        break;
+                    case HashAlgorithmNames.MD5 when Helpers.HasMD5:
+                        // If the platform doesn't support MD5, let it fall in to the default case where
+                        // HashProviderDispenser.CreateHashProvider will throw the appropriate "unknown hash algorithm"
+                        // exception.
+                        modifiedKey = MD5.HashData(key);
+                        break;
+                    default:
```

Comment:
This default case will still be hit when someone calls:

```C#
IncrementalHash.CreateHMAC(new HashAlgorithmName("POTATOALG"), new byte[] { });
```

In Windows, this defers to CNG and CNG gets to decide if it supports the algorithm or not, so in that sense this code path is still useful where .NET today can support some future hash algorithm that CNG learns to support.

For all of the common cases though, we can accelerate them using the one shots.

Reply:
> This default case will still be hit when someone calls:

Actually, that's not true. We pass a block size of `-1` in that case since we never expose the adjusted key there. Realistically, I don't see a way for this `default` case to get hit, so we can change this to just throw an unknown hash algorithm exception. We could keep the code around as a fallback though.

What do you think @bartonjs?


```C#
throw new CryptographicException(SR.Format(SR.Cryptography_UnknownHashAlgorithm, hashAlgorithmId));
```

Reply:
it ends up being the same either way... on Browser CreateHashProvider("MD5") will throw that same exception: https://github.com/dotnet/runtime/blob/7b1fd25ffd29264204e05cd40ce3daabe12a2977/src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/HashProviderDispenser.Browser.cs#L14-L24

So shortcutting it to throw here seems reasonable.